test(e2e): add tests for ES2015 module support

appsforartists opened the pull request:
(Note: this is a continuation of #2685, because @wesleycho had to abandon it.  @dignifiedquire, you already approved that PR.  This is the same one, but rebased off the current HEAD.)

mdrobny commented:
@appsforartists as far as I understand @wesleycho said that he can’t contribute to open source currently so I am not sure if even merging his commit (after your rebase) would work. I think that @wesleycho would be ok with just copying his code and adding as yours for the sake of this project, what do you think guys?

appsforartists commented:
I'd want to offer him proper credit, and I don't think it's legally kosher to claim his work as my own either.

My thinking is that he offered this code before he had any restrictions, so including his commit in this PR should be fine (and I presume he would have closed the earlier PR if it wasn't); however, I don't want to be presumptuous.  @wesleycho, are there any problems with landing code that you wrote before you worked at your current employer?

mattyclarkson commented:
Is there any backup plan if we don't hear from @wesleycho? Really keen to have this!

appsforartists commented:
I chatted with @wesleycho off-thread.  He can't contribute additional code to this PR, but confirmed that his previous PR was licensed to Karma under the CLA.  This can be merged whenever a reviewer gives it the :+1:.

@dignifiedquire, can you please merge?

appsforartists commented:
SGTM

I'm away this week, but will try to get to this next week.

dignifiedquire commented:
@appsforartists I just merged #2846, can you rebase on `master` please?

appsforartists commented:
Rebased.  

e2e tests are taking an unreasonably long time locally, so I haven't been able to verify.  I took @wesleycho's original tests + my patches from the original PR, and changed his cucumber file to use `type: 'module'` instead of `esModule: true`.

appsforartists commented:
Looks like Travis is a lie.  If you [look at the logs](https://s3.amazonaws.com/archive.travis-ci.org/jobs/305422781/log.txt?X-Amz-Expires=30&X-Amz-Date=20171121T212320Z&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAJRYRXRSVGNKPKO5A/20171121/us-east-1/s3/aws4_request&X-Amz-SignedHeaders=host&X-Amz-Signature=3e8f86a6669e5fa48ff79fd3e18cbb47edfc2c108ad51904aa40afd31ec2c660), you'll see

```
Failures:

1) Scenario: [1mSimple middleware[22m - [90mtest/e2e/module-types.feature:6[39m
   Step: [1mThen it passes with:[22m - [90mtest/e2e/module-types.feature:20[39m
   Step Definition: [90mtest/e2e/step_definitions/core_steps.js:152[39m
   Message:
     [31mError: Expected output to match the following:
       ..
     Firefox
     Got:
       Firefox 57.0.0 (Linux 0.0.0): Executed 0 of 0 ERROR (0.001 secs / 0 secs)
     [39m[90m
         at World.<anonymous> (/home/travis/build/karma-runner/karma/test/e2e/step_definitions/core_steps.js:173:23)
     [39m

45 scenarios ([31m1 failed[39m, [32m44 passed[39m)
155 steps ([31m1 failed[39m, [32m154 passed[39m)
1m17.262s
[31m>> [39mfailed tests, please see the output
```

so it shouldn't have the 💚 .

dignifiedquire commented:
Spoke too soon, the test seems to be failing, is it working for you locally?

appsforartists commented:
@dignifiedquire 

Took me a bit to figure out how to get the right test to run against the built karma, but I've got that sorted now.

All the right files are being loaded and code is being executed, but the closures being passed into `if` calls aren't being called.

With these logs:
```typescript
import { plus } from '../plus.js'

console.log('a')

describe('plus', function () {
  console.log('b')

  it('should pass', function () {
    console.log('c')
    expect(true).toBe(true)
  })

  it('should work', function () {
    expect(plus(1, 2)).toBe(3)
  })
})
```
You'd see `a` and `b`, but not `c`.  `describe` and `it` are defined.  Any idea what's going on?

appsforartists commented:
I tried using the `basic` tests, and it appears `type: 'module'` breaks `jasmine`'s ability to identify tests.

This works:
```
      files = [
        { pattern: 'basic/plus.js' },
        { pattern: 'basic/test.js' },
      ];
```

but this fails:
```
      files = [
        { pattern: 'basic/plus.js', type: 'module' },
        { pattern: 'basic/test.js', type: 'module' },
      ];
```

appsforartists commented:
I just tried `mocha` as well and had the same problem.  Current hypothesis: the `mocha` and `jasmine` plugins somehow wrap the JS test files, and `type: 'module'` is being excluded.

appsforartists commented:
@dignifiedquire I'm not sure what the purpose of `expect(true).to.be.true` is in each of these tests, but it seems that other e2e karma suites have them too.

I think you should merge this, even though the tests don't pass.  Presumably if they don't pass in cucumber, #2846 doesn't work for Karma users either.  We should have a test to demonstrate this breakage and determine when it's fixed.

Coridyn commented:
@appsforartists I've had a look and worked out why `jasmine` isn't running tests from modules.

[ES modules run as deferred](https://jakearchibald.com/2017/es-modules-in-browsers/#defer-by-default) but karma kicks off the tests in a synchronous script.

This means that karma starts running tests before the modules have been loaded and therefor those tests don't run.

A possible fix for this would be to change the `window.__karma__.loaded()` call to happen in its own module or in the `DOMContentLoaded` event, after the ES modules have been loaded.

This change would need to be made in [`static/context.html`](https://github.com/karma-runner/karma/blob/b53929ad7a4cd0fcac41b36ae28ca44a2d2e930c/static/context.html#L30) and [`static/debug.html`](https://github.com/karma-runner/karma/blob/b53929ad7a4cd0fcac41b36ae28ca44a2d2e930c/static/debug.html#L32).

Current code:
```html
  <script type="text/javascript">
    window.__karma__.loaded();
  </script>
```


Proposed change:
```html
  <script type="text/javascript" nomodule>
    window.__karma__.loaded();
  </script>
  <script type="module">
    window.__karma__.loaded();
  </script>
```

Or alternatively, using the `DOMContentLoaded` event:

```html
  <script type="text/javascript">
    document.addEventListener("DOMContentLoaded", function() {
      window.__karma__.loaded();
    });
  </script>
```

A similar change would need to be done by anyone who is using a custom `context.html` file.

I'm not sure if this would affect the `static/client_with_context.html` file as I don't know how the tests are kicked off in that scenario.

appsforartists commented:
Thanks @Coridyn! (You might have noticed this is the first time I've PRed karma; not terribly familiar with its internals.)

I've made those changes.  The tests are passing locally; hopefully in 20m, they'll pass on Travis too.

appsforartists commented:
HOORAY!  Travis passed.  (Took more work than I expected, because apparently Chrome doesn't work on Karma's Travis setup, and Firefox doesn't enable modules by default.)

lastmjs commented:
Anything I can help with to move this along? I could really use these changes

johnjbarton commented:
@lastmjs you can figure out why the appveyor build fails.

surma commented:
As a workaround for people like me who are waiting for this to land upstream: You can rip out `static/context.html` from this PR and put it in your `karma.conf.js` 

```js
const configuration = {
  /* ... */
  customContextFile: 'tests/context.html',
  /* ... */
};
```

surma commented:
Another note: The provided `context.html` will just ignore test files that don’t load successfully instead of failing the test suite.

Example:

```js
// module.test.js — loaded as a module
import * as Something from '/this-file-doesnt-exist.js';

describe('Something', function () {
  it('should do something', function () {
    // ...
  });
});
```

This test file should fail as the import does not exist. Instead Karma just ignores the file and succeeds.

jehon commented:
I tried to fix this PR, made a checkout etc... but could not make it run.

So, because I am a bit impatient, I made a small plugin here: https://www.npmjs.com/package/karma-jasmine-es6. This plugin took me 30 minutes, so it is not really a waste of time.

This plugin is meant to be a temporary fix for a short time, while waiting for this PR to accomplish its mission.

Btw, thanks in advance to anybody who would manage to fix this PR !

vitalets commented:
Hi everyone! Thanks to this issue - I've got ES modules working for local files. But it throws error for imports from `node_modules`. For example: 
```js
// test.js
import uuid from 'uuid';
```
Outputs:
```
Chrome 68.0.3440 (Mac OS X 10.13.6) ERROR
  {
    "message": "Uncaught TypeError: Failed to resolve module specifier \"uuid\". Relative references must start with either \"/\", \"./\", or \"../\".\nat http://localhost:9876/context.html:0:0\n\nTypeError: Failed to resolve module specifier \"uuid\". Relative references must start with either \"/\", \"./\", or \"../\".",
    "str": "Uncaught TypeError: Failed to resolve module specifier \"uuid\". Relative references must start with either \"/\", \"./\", or \"../\".\nat http://localhost:9876/context.html:0:0\n\nTypeError: Failed to resolve module specifier \"uuid\". Relative references must start with either \"/\", \"./\", or \"../\"."
  }
```

How do you solve this?
Thanks!

appsforartists commented:
@vitalets Node's module resolution is supported by Webpack, but not part of the ES2015 spec, which means browsers don't support it.

You can either make your imports paths, e.g.:

```typescript
import {} from './node_modules/some_library/dist/some_file.js';
```

or run your tests against a bundled source, e.g. from Webpack.

johnjbarton commented:
Fixed by #3072
